File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    private Set<Integer> getPrimes(int acc, List<Integer> numbers){
        //종료조건
        Set<Integer> primes = new HashSet<>();
        if(isPrime(acc)) primes.add(acc);
        if(numbers.isEmpty()) return primes;
        
        //상태전이 구현
        for(int i = 0; i < numbers.size(); i++){
            //numbers.get(i)로 상태 전이 진행
            int nextAcc = acc * 10 + numbers.get(i);
            List<Integer> nextNumbers = new ArrayList<>(numbers);
            nextNumbers.remove(i);
            primes.addAll(getPrimes(nextAcc, nextNumbers));
        }
        
        return primes;
    }
    
    private boolean isPrime(int n){
        if(n <= 1) return false;
        //n의 제곱근 보다 작은 수 들을 체크한다
        for(int i = 2; i * i <= n; i++){
            if(n % i == 0) return false;
        }
        return true;
    }
    public int solution(String nums) {
        List<Integer> numbers = nums.chars()
            .map(c -> c - '0')
            .boxed()
            .collect(Collectors.toList());
        
        return getPrimes(0, numbers).size();
    }
}
/*
상태
1. 지금까지 만들어 놓은 숫자
2. 사용할 수 있는 종이 조각들
(acc, number)
*/

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations

def isprime(n):
    if n < 2: 
        return 0
    for i in range(2, int(n ** 0.5)+1):
        if n % i == 0:
            return 0
    return 1
        
def solution(numbers):
    numbers = list(numbers);
    num = []
    for i in range(1, len(numbers)+1):
        num.append(permutations(numbers, i))
    num = set([int(''.join(y)) for x in num for y in x])
    return len([n for n in num if isprime(n)])
    # ret = set(n for num in n if isprime(n))
    # return len(ret)
    
"""
1. 뽑는 갯수마다 순열 구해서 숫자를 저장
2. 숫자가 소수인지 확인해서 갯수반환

"""